README.md
# jscrypto
[![npm version](https://badge.fury.io/js/jscrypto.svg)](https://badge.fury.io/js/jscrypto)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
![Build](https://github.com/Hinaser/jscrypto/actions/workflows/test.yml/badge.svg?)

[crypto-js](https://github.com/brix/crypto-js) enhancement for modern js environments.

- Works in modern browsers and **IE9/10/11**.  
  \*IE9/10 uses weak random generator on cipher encryption with string password. Use it at your own risk.  
  \*If only using decryption or hash/hmac, weak random generator does not cause any trouble.
- Loadable with ES6/CommonJS/Typescript/Browser runtimes.
- Support [`GCM`][GCM]/[`GMAC`][GMAC]/[`CCM`][CCM]/[`CBC-MAC`][CBC-MAC] :tada::tada::tada:
- [CLI commands](#cli) available:  
  i.e. `npx jscrypto sha1 "message"`, `npx jscrypto aes enc "message" "password"`, etc.  
- Written in Typescript with rich type declarations.
- When bundling only SHA256 module, the webpack-ed js file can be less than 6kb.  
- Default parameters for Block cipher (AES/DES/Triple-DES) is tuned to be OpenSSL(1.1.1f) compatible.  
  Read further [here](#openssl-compat)

### Breaking changes for version 0.x.x users
There are several breaking changes between version 0.x and 1.x.  
Please see detail in [CHANGELOG](https://github.com/Hinaser/jscrypto/blob/master/CHANGELOG.md)


## Install

```
npm install jscrypto
# or
yarn add jscrypto
```

If you only want to use [CLI commands](#cli), you don't even need to install `jscrypto`.  
Just dispatch `npx` command like `npx jscrypto sha256 "message"`.  
Read further [here](#cli)

## Usage
### CommonJS Environment (Node.js environment like node CLI, AWS Lambda, etc)
```js
// Load whole library modules.
const JsCrypto = require("jscrypto");
console.log(JsCrypto.SHA256.hash("test").toString());

// or load only necessary modules (Recommended for faster loading and reduced size)
const {SHA256} = require("jscrypto/SHA256");
console.log(SHA256.hash("test").toString());
```

### ES6 Environment (i.e. Creating app/library with webpack/react-scripts or some es6-compatible bundlers)
**Be sure to load the module from `jscrypto/es6`.**  
This can greatly reduce bundle size by bundlers tree-shaking ability. 
Don't forget to add `/es6` following `jscrypto`
```ecmascript 6
// Load whole library modules.
import JsCrypto from "jscrypto/es6";
console.log(JsCrypto.SHA256.hash("test").toString());
...
import {SHA256} from "jscrypto/es6/SHA256"; // Recommended
console.log(SHA256.hash("test").toString());
```

### Typescript Environment
**Be sure to load the module from `jscrypto/es6`.**
```ecmascript 6
// Load whole library modules.
import * as JsCrypto from "jscrypto/es6";
console.log(JsCrypto.SHA256.hash("test").toString());
...
import {SHA256} from "jscrypto/es6/SHA256"; // Recommended
console.log(SHA256.hash("test").toString());
```

### Browser
Copy js files/directories under `/dist` dir into somewhere browser can access.  
Then directly load js file into `<script>` tag.
```html
<script src="dist/index.js"></script> <!-- All modules are loaded into browser -->
<script type="text/javascript">
  // This will output: "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"
  console.log(JsCrypto.SHA256.hash("test").toString());
</script>
<!-- OR -->
<script src="dist/SHA256.js"></script> <!-- Single module loading is lightweight and faster. -->
<script type="text/javascript">
  // This will output: "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"
  console.log(JsCrypto.SHA256.hash("test").toString());
</script>
```

<h3 id='cli'>CLI</h3>

Command Line Interface to try various crypto modules on terminal.
```cmd
  Usage: npx jscrypto <hash|hmac|cipher> [command options]

  hash: md5, sha1, sha3, sha224, sha256, sha384, sha512, ripemd160
  hmac: hmac-md5, hmac-sha1, hmac-sha224, hmac-sha256, hmac-sha384, hmac-sha512
  cipher: aes, des, des3, rc4

  
  $ npx jscrypto <hash> message [-msg hex|base64|utf8] [-out hex|base64]

    default:
      -msg: utf8 ... recognize message as utf-8 string
      -out: hex ... output hashed binary as hex string
    example:
      #Output of below 3 examples are the same
      npx jscrypto sha1 test
      npx jscrypto sha1 74657374 -msg hex
      npx jscrypto sha1 dGVzdA== -msg base64


  $ npx jscrypto <hmac> message key [msg hex|base64|utf8] [-key hex|base64|utf8] [-out hex|base64]

    default:
      -msg: utf8 ... recognize message as utf-8 string
      -key: utf8 ... recognize key as utf-8 string
      -out: hex ... output hashed binary as hex string
    example:
      #Output of below 3 examples are the same
      npx jscrypto hmac-sha1 test key
      npx jscrypto hmac-sha1 74657374 6b6579 -msg hex -key hex
      npx jscrypto hmac-sha1 dGVzdA== a2V5 -msg base64 -key base64


  $ npx jscrypto <cipher> message key [-msg hex|base64|utf8] [-key hex|base64|utf8] [-out hex|base64|utf8] [-mode cbc|ecb|ofb|cfb] [-pad pkcs7|iso10126|iso97971|ansix923|nopadding] [-kdf pbkdf2|evpkdf]

    default:
      -msg: utf8 ... recognize message as utf-8 string
      -key: utf8 ... recognize key as utf-8 string
      -out: base64|hex ... base64 on encryption, hex on decryption. Note: utf8 cannot be used on encryption.
      -mode: cbc ... Code block chaining as block cipher mode
      -pad: pkcs7 ... Pkcs7 padding as block padding
      -kdf: pbkdf2 ... PBKDF2 as key derivation function
    example:
      #Encrypt (Output would not be the same because of a random salt, but can be decrypted with the same key)
      npx jscrypto aes enc test password
      npx jscrypto aes enc 74657374 70617373776f7264 -msg hex -key hex
      npx jscrypto aes enc dGVzdA== cGFzc3dvcmQ= -msg base64 -key base64
      #Decrypt
      npx jscrypto aes dec U2FsdGVkX19Kf/wItWMuaTrQYV3OljA3Cr9WPMhC6Tk= password -out utf8
      npx jscrypto aes dec A2pYDd/3oeENsRFGA1Y0Mg== 70617373776f7264 -key hex -out utf8
      npx jscrypto aes dec A2pYDd/3oeENsRFGA1Y0Mg== cGFzc3dvcmQ= -key base64 -out utf8
```

<h3 id='openssl-compat'>OpenSSL compatibility</h3>

#### Encryption
  ```js
  encryptedData = JsCrypto.AES.encrypt("message", "secret phrase").toString();
  ```
is equivalent in OpenSSL (1.1.1f) to
  ```shell
  echo -n "message" | openssl enc -e -aes-256-cbc -pass pass:"secret phrase" -base64 -pbkdf2
  # Note: Because of a random salt, everytime it produces different base64 string.
  # But it is OK for decryption.
  ```

#### Decryption
Encrypted data can be decrypted by
  ```js
  JsCrypto.AES.decrypt(encryptedData, "secret phrase").toString(JsCrypto.Utf8); // "message"
  ```
or in OpenSSL
  ```shell
  echo "U2FsdGVkX1..." | openssl enc -d -aes-256-cbc -pass pass:"secret phrase" -base64 -pbkdf2
  # U2FsdGVkX1... is the output from either JsCrypto/OpenSSL encryption code/command.
  ```

## FAQ
#### Failed to import jscrypto in Typescript environment.

In most cases, your `tsconfig.json` is configured not to load npm module from `node_modules` folder.  
Check your `tsconfig.js` to be:
```json
{
  "compilerOptions": {
    "moduleResolution": "Node"
  }
}
```

## API
`jscrypto` supports crypto modules as well as `cryptojs`.

### *Popular*
**Hash** [`MD5`][MD5], [`SHA1`][SHA1], [`SHA3`][SHA3], [`SHA224`][SHA224], [`SHA256`][SHA256], [`SHA384`][SHA384], [`SHA512`][SHA512], [`RIPEMD160`][RIPEMD160],  
**Message/Key Hash** [`HMAC-MD5`][HMAC-MD5], [`HMAC-SHA224`][HMAC-SHA224], [`HMAC-SHA256`][HMAC-SHA256], [`HMAC-SHA384`][HMAC-SHA384], [`HMAC-SHA512`][HMAC-SHA512], [`GMAC`][GMAC], [`CBC-MAC`][CBC-MAC]  
**Block Cipher** [`AES`][AES], [`DES`][DES], [`Triple-DES`][Triple-DES]

### *Basic structure*
**Word** [`Word32Array`][Word32Array], [`Word64Array`][Word64Array]  
**Encoder** [`Base64`][Base64], [`Hex`][Hex], [`Latin1`][Latin1], [`Utf8`][Utf8], [`Utf16`][Utf16]

### *Misc*
**Stream Cipher** [`Rabbits`][Rabbits], [`RC4`][RC4], [`RC4Drop`][RC4Drop]  
**Key Derive Function** [`OpenSSLKDF`][OpenSSLKDF], [`EvpKDF`][EvpKDF], [`PBKDF2`][PBKDF2]  
**Block Cipher mode** [`CBC`][CBC], [`CFB`][CFB], [`CTR`][CTR], [`ECB`][ECB], [`OFB`][OFB], [`GCM`][GCM], [`CCM`][CCM]  
**Block Padding** [`AnsiX923`][AnsiX923], [`ISO10126`][ISO10126], [`ISO97971`][ISO97971], [`NoPadding`][NoPadding], [`Pkcs7`][Pkcs7], [`Zero`][Zero]  
**Formatter** [`OpenSSLFormatter`][OpenSSLFormatter]

[MD5]: https://github.com/Hinaser/jscrypto/blob/master/API.md#md5
[SHA1]: https://github.com/Hinaser/jscrypto/blob/master/API.md#sha1
[SHA3]: https://github.com/Hinaser/jscrypto/blob/master/API.md#sha3
[SHA224]: https://github.com/Hinaser/jscrypto/blob/master/API.md#sha224
[SHA256]: https://github.com/Hinaser/jscrypto/blob/master/API.md#sha256
[SHA384]: https://github.com/Hinaser/jscrypto/blob/master/API.md#sha384
[SHA512]: https://github.com/Hinaser/jscrypto/blob/master/API.md#sha512
[RIPEMD160]: https://github.com/Hinaser/jscrypto/blob/master/API.md#ripemd160
[HMAC-MD5]: https://github.com/Hinaser/jscrypto/blob/master/API.md#hmac-md5
[HMAC-SHA224]: https://github.com/Hinaser/jscrypto/blob/master/API.md#hmac-sha224
[HMAC-SHA256]: https://github.com/Hinaser/jscrypto/blob/master/API.md#hmac-sha256
[HMAC-SHA384]: https://github.com/Hinaser/jscrypto/blob/master/API.md#hmac-sha384
[HMAC-SHA512]: https://github.com/Hinaser/jscrypto/blob/master/API.md#hmac-sha512
[GMAC]: https://github.com/Hinaser/jscrypto/blob/master/API.md#gmac
[CBC-MAC]: https://github.com/Hinaser/jscrypto/blob/master/API.md#cbc-mac
[AES]: https://github.com/Hinaser/jscrypto/blob/master/API.md#aes
[AES-GCM]: https://github.com/Hinaser/jscrypto/blob/master/API.md#aes-gcm
[AES-CCM]: https://github.com/Hinaser/jscrypto/blob/master/API.md#aes-ccm
[DES]: https://github.com/Hinaser/jscrypto/blob/master/API.md#des
[Triple-DES]: https://github.com/Hinaser/jscrypto/blob/master/API.md#des3
[Word32Array]: https://github.com/Hinaser/jscrypto/blob/master/API.md#word32array
[Word64Array]: https://github.com/Hinaser/jscrypto/blob/master/API.md#mword64Array
[Base64]: https://github.com/Hinaser/jscrypto/blob/master/API.md#base64
[Hex]: https://github.com/Hinaser/jscrypto/blob/master/API.md#hex
[Latin1]: https://github.com/Hinaser/jscrypto/blob/master/API.md#latin1
[Utf8]: https://github.com/Hinaser/jscrypto/blob/master/API.md#utf8
[Utf16]: https://github.com/Hinaser/jscrypto/blob/master/API.md#utf16
[Rabbits]: https://github.com/Hinaser/jscrypto/blob/master/API.md#rabbits
[RC4]: https://github.com/Hinaser/jscrypto/blob/master/API.md#rc4
[RC4Drop]: https://github.com/Hinaser/jscrypto/blob/master/API.md#rc4drop
[OpenSSLKDF]: https://github.com/Hinaser/jscrypto/blob/master/API.md#opensslkdf
[EvpKDF]: https://github.com/Hinaser/jscrypto/blob/master/API.md#evpkdf
[PBKDF2]: https://github.com/Hinaser/jscrypto/blob/master/API.md#pbkdf2
[CBC]: https://github.com/Hinaser/jscrypto/blob/master/API.md#cbc
[CFB]: https://github.com/Hinaser/jscrypto/blob/master/API.md#cfb
[CTR]: https://github.com/Hinaser/jscrypto/blob/master/API.md#ctr
[ECB]: https://github.com/Hinaser/jscrypto/blob/master/API.md#ecb
[OFB]: https://github.com/Hinaser/jscrypto/blob/master/API.md#ofb
[GCM]: https://github.com/Hinaser/jscrypto/blob/master/API.md#gcm
[CCM]: https://github.com/Hinaser/jscrypto/blob/master/API.md#ccm
[AnsiX923]: https://github.com/Hinaser/jscrypto/blob/master/API.md#ansix923
[ISO10126]: https://github.com/Hinaser/jscrypto/blob/master/API.md#iso10126
[ISO97971]: https://github.com/Hinaser/jscrypto/blob/master/API.md#iso97971
[NoPadding]: https://github.com/Hinaser/jscrypto/blob/master/API.md#nopadding
[Pkcs7]: https://github.com/Hinaser/jscrypto/blob/master/API.md#pkcs7
[Zero]: https://github.com/Hinaser/jscrypto/blob/master/API.md#zero
[OpenSSLFormatter]: https://github.com/Hinaser/jscrypto/blob/master/API.md#opensslformatter


CHANGELOG.md
# Changelog

## [1.0.3] - 2022-01-12
### Fixed
- Fixed an issue where AES-CCM got wrong MAC while payload padding required

## [1.0.2] - 2021-07-08
### Fixed
- Fixed an issue where invalid Hex string could be parsed without error.  
  For example, invalid Hex string `1g` was recognized as `0x01`.

## [1.0.1] - 2021-06-20
### Fixed
- Fixed an issue where building jscrypto with webpack shows warning messages like below:  
  `Critical dependency: the request of a dependency is an expression`

## [1.0.0] - 2021-04-22
### Added
- Added [CCM block cipher mode (Counter mode/CBC-MAC)](https://github.com/Hinaser/jscrypto/blob/master/API.md#ccm)  
  https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38c.pdf
- Added [CBC-MAC](https://github.com/Hinaser/jscrypto/blob/master/API.md#cbc-mac)
- Added tag length option to [GMAC](https://github.com/Hinaser/jscrypto/blob/master/API.md#gmac) and [GCM](https://github.com/Hinaser/jscrypto/blob/master/API.md#gcm)
- Exposed `CipherParams` class to public.

### Changed
**Small breaking change**
- Pass initializing vector as Word32Array to constructor of BlockCipherMode instead of just a 32bit number array.  
  \*This may enable developers to use non-32bit-aligned iv value to block cipher mode in future release.  
  \*This change may not be breaking lib compatibility unless developers directly instantiate BlockCipherMode or  
    creating original BlockCipherMode extending old BlockCipherMode class.
  ```js
  // BEFORE
  const gcm = new GCM({cipher: AES, iv: [0x11223344]});
  
  // AFTER
  const gcm = new GCM({cipher: AES, iv: new Word32Array([0x11223344], 4)});
  ```

**Breaking change**
- Changed GMAC hash function name from `GCM.hash()` to `GCM.mac()`.
- Calculating authTag in GCM now requires developer to manually call authTag function.  
  Encryption/Decryption and MAC Generation are now calculated independently.
  ```js
  //////////////////////
  // AES-GCM
  //////////////////////
  // BEFORE
  const encrypted = AES.encrypt(msg, key, { iv, mode: GCM, padding: NoPadding, authData });
  encrypted.authTag !== undefined; // This returns true. authTag is automatically calculated on encryption.

  // AFTER
  const encrypted = AES.encrypt(msg, key, { iv, mode: GCM, padding: NoPadding, authData });
  encrypted.authTag === undefined; // This returns true. authTag must be manually calculated as below.
  const authTag = GCM.mac(AES, key, iv, authData, encrypted.cipherText);
  ```

## [0.2.0] - 2021-04-07
### Added
- Added GCM block cipher mode. (Galois Counter Mode)
- Added GMAC (Galois Message Authentication Code)

## [0.1.0] - 2021-03-30
### Added
- Added CLI executable  
- Added HmacSHA1 (`jscrypto/HmacSHA1`, `jscrypto/es6/HmacSHA1`)

## [0.0.2] - 2021-03-28
### Changed
- Removed typescript source file from npm package to reduce package size.
- Word32Array can be initialized/cloned by `new Word32Array(anotherWord32Array)`.

### Fixed
- Added missing kdf modules(`jscrypto/EvpKDF`, `jscrypto/PBKDF2`, `jscrypto/OpenSSLKDF`).

## [0.0.1] - 2021-03-23
### Fixed
- Fixed an issue where encoder/kdf modules have the different loading path.  
  ```html
  <!-- BEFORE THIS FIX -->
  <!-- Load from index.js -->
  <script src="dist/index.js"></script>
  var word = JsCrypto.Hex.parse("00112233");
  <!-- Load from individual modules -->
  <script src="dist/encoder/Hex.js"></script>
  var word = JsCrypto.encode.Hex.parse("00112233");
  
  <!-- AFTER THIS FIX -->
  <script src="dist/Hex.js"></script>
  var word = JsCrypto.Hex.parse("00112233");
  ```
- Fixed an issue where loading multiple modules in browser only preserves the last module.
   ```html
  <script src="dist/AES.js"></script>
  <script src="dist/DES.js"></script>
  // OK
  var word1 = JsCrypto.DES.encrypt("message", "key");
  // ERROR
  var word2 = JsCrypto.AES.encrypt("message", "key");
   ```

## [0.0.0] - 2021-03-21
Initial release.

<!-- [Unreleased]: https://github.com/Hinaser/jscrypto/compare/v0.1.0...v0.2.0 -->
[1.0.3]: https://github.com/Hinaser/jscrypto/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/Hinaser/jscrypto/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/Hinaser/jscrypto/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/Hinaser/jscrypto/compare/v0.2.0...v1.0.0
[0.2.0]: https://github.com/Hinaser/jscrypto/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/Hinaser/jscrypto/compare/v0.0.2...v0.1.0
[0.0.2]: https://github.com/Hinaser/jscrypto/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/Hinaser/jscrypto/compare/v0.0.0...v0.0.1
[0.0.0]: https://github.com/Hinaser/jscrypto/compare/tag/v0.0.0

src/MD5.ts
import {Word32Array} from "./lib/Word32Array";
import {Hasher, HasherProps} from "./lib/algorithm/Hasher";

export interface MD5Props extends HasherProps {
  hash: Word32Array;
}

// Constants table
const T: number[] = [];

(function computeConstant(){
  for(let i=0;i<64;i++){
    T[i] = (Math.abs(Math.sin(i+1)) * 0x100000000) | 0;
  }
})();

function FF(a: number, b: number, c: number, d: number, x: number, s: number, t: number){
  const n = a + ((b & c) | (~b & d)) + x + t;
  return ((n << s) | (n >>> (32 - s))) + b;
}

function GG(a: number, b: number, c: number, d: number, x: number, s: number, t: number){
  const n = a + ((b & d) | (c & ~d)) + x + t;
  return ((n << s) | (n >>> (32 - s))) + b;
}

function HH(a: number, b: number, c: number, d: number, x: number, s: number, t: number){
  const n = a + (b ^ c ^ d) + x + t;
  return ((n << s) | (n >>> (32 - s))) + b;
}

function II(a: number, b: number, c: number, d: number, x: number, s: number, t: number){
  const n = a + (c ^ (b | ~d)) + x + t;
  return ((n << s) | (n >>> (32 - s))) + b;
}

/**
 * MD5 hash algorithm
 */
export class MD5 extends Hasher {
  private _hash: Word32Array = new Word32Array([
    0x67452301, 0xefcdab89,
    0x98badcfe, 0x10325476
  ]);
  
  public constructor(props?: MD5Props) {
    super(props);
    if(props && typeof props.hash !== "undefined"){
      this._hash = props.hash.clone();
    }
  }
  
  protected _doReset() {
    this._hash = new Word32Array([
      0x67452301, 0xefcdab89,
      0x98badcfe, 0x10325476
    ]);
  }
  
  protected _doProcessBlock(words: number[], offset: number) {
    // Swap endian
    for (let i=0;i<16;i++) {
      // Shortcuts
      const offsetI = offset + i;
      const wordsOffsetI = words[offsetI];
    
      words[offsetI] = (
        (((wordsOffsetI << 8)  | (wordsOffsetI >>> 24)) & 0x00ff00ff)
        | (((wordsOffsetI << 24) | (wordsOffsetI >>> 8))  & 0xff00ff00)
      );
    }
  
    // Shortcuts
    const H = this._hash.words;
  
    const wordOffset0  = words[offset];
    const wordOffset1  = words[offset + 1];
    const wordOffset2  = words[offset + 2];
    const wordOffset3  = words[offset + 3];
    const wordOffset4  = words[offset + 4];
    const wordOffset5  = words[offset + 5];
    const wordOffset6  = words[offset + 6];
    const wordOffset7  = words[offset + 7];
    const wordOffset8  = words[offset + 8];
    const wordOffset9  = words[offset + 9];
    const wordOffset10 = words[offset + 10];
    const wordOffset11 = words[offset + 11];
    const wordOffset12 = words[offset + 12];
    const wordOffset13 = words[offset + 13];
    const wordOffset14 = words[offset + 14];
    const wordOffset15 = words[offset + 15];
  
    // Working variables
    let a = H[0];
    let b = H[1];
    let c = H[2];
    let d = H[3];
  
    // Computation
    a = FF(a, b, c, d, wordOffset0,  7,  T[0]);
    d = FF(d, a, b, c, wordOffset1,  12, T[1]);
    c = FF(c, d, a, b, wordOffset2,  17, T[2]);
    b = FF(b, c, d, a, wordOffset3,  22, T[3]);
    a = FF(a, b, c, d, wordOffset4,  7,  T[4]);
    d = FF(d, a, b, c, wordOffset5,  12, T[5]);
    c = FF(c, d, a, b, wordOffset6,  17, T[6]);
    b = FF(b, c, d, a, wordOffset7,  22, T[7]);
    a = FF(a, b, c, d, wordOffset8,  7,  T[8]);
    d = FF(d, a, b, c, wordOffset9,  12, T[9]);
    c = FF(c, d, a, b, wordOffset10, 17, T[10]);
    b = FF(b, c, d, a, wordOffset11, 22, T[11]);
    a = FF(a, b, c, d, wordOffset12, 7,  T[12]);
    d = FF(d, a, b, c, wordOffset13, 12, T[13]);
    c = FF(c, d, a, b, wordOffset14, 17, T[14]);
    b = FF(b, c, d, a, wordOffset15, 22, T[15]);
  
    a = GG(a, b, c, d, wordOffset1,  5,  T[16]);
    d = GG(d, a, b, c, wordOffset6,  9,  T[17]);
    c = GG(c, d, a, b, wordOffset11, 14, T[18]);
    b = GG(b, c, d, a, wordOffset0,  20, T[19]);
    a = GG(a, b, c, d, wordOffset5,  5,  T[20]);
    d = GG(d, a, b, c, wordOffset10, 9,  T[21]);
    c = GG(c, d, a, b, wordOffset15, 14, T[22]);
    b = GG(b, c, d, a, wordOffset4,  20, T[23]);
    a = GG(a, b, c, d, wordOffset9,  5,  T[24]);
    d = GG(d, a, b, c, wordOffset14, 9,  T[25]);
    c = GG(c, d, a, b, wordOffset3,  14, T[26]);
    b = GG(b, c, d, a, wordOffset8,  20, T[27]);
    a = GG(a, b, c, d, wordOffset13, 5,  T[28]);
    d = GG(d, a, b, c, wordOffset2,  9,  T[29]);
    c = GG(c, d, a, b, wordOffset7,  14, T[30]);
    b = GG(b, c, d, a, wordOffset12, 20, T[31]);
  
    a = HH(a, b, c, d, wordOffset5,  4,  T[32]);
    d = HH(d, a, b, c, wordOffset8,  11, T[33]);
    c = HH(c, d, a, b, wordOffset11, 16, T[34]);
    b = HH(b, c, d, a, wordOffset14, 23, T[35]);
    a = HH(a, b, c, d, wordOffset1,  4,  T[36]);
    d = HH(d, a, b, c, wordOffset4,  11, T[37]);
    c = HH(c, d, a, b, wordOffset7,  16, T[38]);
    b = HH(b, c, d, a, wordOffset10, 23, T[39]);
    a = HH(a, b, c, d, wordOffset13, 4,  T[40]);
    d = HH(d, a, b, c, wordOffset0,  11, T[41]);
    c = HH(c, d, a, b, wordOffset3,  16, T[42]);
    b = HH(b, c, d, a, wordOffset6,  23, T[43]);
    a = HH(a, b, c, d, wordOffset9,  4,  T[44]);
    d = HH(d, a, b, c, wordOffset12, 11, T[45]);
    c = HH(c, d, a, b, wordOffset15, 16, T[46]);
    b = HH(b, c, d, a, wordOffset2,  23, T[47]);
  
    a = II(a, b, c, d, wordOffset0,  6,  T[48]);
    d = II(d, a, b, c, wordOffset7,  10, T[49]);
    c = II(c, d, a, b, wordOffset14, 15, T[50]);
    b = II(b, c, d, a, wordOffset5,  21, T[51]);
    a = II(a, b, c, d, wordOffset12, 6,  T[52]);
    d = II(d, a, b, c, wordOffset3,  10, T[53]);
    c = II(c, d, a, b, wordOffset10, 15, T[54]);
    b = II(b, c, d, a, wordOffset1,  21, T[55]);
    a = II(a, b, c, d, wordOffset8,  6,  T[56]);
    d = II(d, a, b, c, wordOffset15, 10, T[57]);
    c = II(c, d, a, b, wordOffset6,  15, T[58]);
    b = II(b, c, d, a, wordOffset13, 21, T[59]);
    a = II(a, b, c, d, wordOffset4,  6,  T[60]);
    d = II(d, a, b, c, wordOffset11, 10, T[61]);
    c = II(c, d, a, b, wordOffset2,  15, T[62]);
    b = II(b, c, d, a, wordOffset9,  21, T[63]);
  
    // Intermediate hash value
    H[0] = (H[0] + a) | 0;
    H[1] = (H[1] + b) | 0;
    H[2] = (H[2] + c) | 0;
    H[3] = (H[3] + d) | 0;
  }
  
  protected _doFinalize(): Word32Array {
    // Shortcuts
    const data = this._data;
    const dataWords = data.words;
  
    const nBitsTotal = this._nBytes * 8;
    const nBitsLeft = data.nSigBytes * 8;
  
    // Add padding
    dataWords[nBitsLeft >>> 5] |= 0x80 << (24 - nBitsLeft % 32);
  
    const nBitsTotalH = Math.floor(nBitsTotal / 0x100000000);
    const nBitsTotalL = nBitsTotal;
    dataWords[(((nBitsLeft + 64) >>> 9) << 4) + 15] = (
      (((nBitsTotalH << 8)  | (nBitsTotalH >>> 24)) & 0x00ff00ff) |
      (((nBitsTotalH << 24) | (nBitsTotalH >>> 8))  & 0xff00ff00)
    );
    dataWords[(((nBitsLeft + 64) >>> 9) << 4) + 14] = (
      (((nBitsTotalL << 8)  | (nBitsTotalL >>> 24)) & 0x00ff00ff) |
      (((nBitsTotalL << 24) | (nBitsTotalL >>> 8))  & 0xff00ff00)
    );
  
    data.nSigBytes = (dataWords.length + 1) * 4;
  
    // Hash final blocks
    this._process();
  
    // Shortcuts
    const hash = this._hash;
    const H = hash.words;
  
    // Swap endian
    for (let i=0;i<4;i++) {
      // Shortcut
      const Hi = H[i];
      H[i] = (((Hi << 8)  | (Hi >>> 24)) & 0x00ff00ff)
        | (((Hi << 24) | (Hi >>> 8))  & 0xff00ff00);
    }
  
    // Return final computed hash
    return hash;
  }
  
  public clone(){
    const props = {hash: this._hash, blockSize: this._blockSize, data: this._data, nBytes: this._nBytes};
    return new MD5(props);
  }
  
  public static hash(message: Word32Array|string){
    return new MD5().finalize(message);
  }
}
